blackbird_python/blackbird/listener.py

```diff
@@ -324,11 +369,59 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
                         # Therefore, it includes a regref statement.
                         op_kwargs[k] = RegRefTransform(v)
 
-            self._program._operations.append(
-                {"op": op, "args": op_args, "kwargs": op_kwargs, "modes": modes}
-            )
+            operation = {"op": op, "args": op_args, "kwargs": op_kwargs, "modes": modes}
+        else:
+            operation = {"op": op, "modes": modes}
+
+        # check if statement is included from another file
+        if operation["op"] in self._includes:
+            bb = self._includes[operation["op"]][1]
+
+            # make sure modes match
+            if len(operation["modes"]) != len(bb.modes):
+                raise ValueError(
+                    "Included operation {} acts on {} modes, "
+                    "but {} modes provided".format(
+                        operation["op"], len(bb.modes), len(operation["modes"])
+                    )
+                )
+
+            if "kwargs" in operation:
+                # operation is a template
+                if not bb.is_template():
+                    raise ValueError(
+                        "Included operation {} does not accept arguments".format(operation["op"])
+                    )
+
+                if bb.parameters != set(operation["kwargs"]):
+                    raise ValueError(
+                        "Included operation {} must accept only keyword arguments {}".format(
+                            operation["op"], bb.parameters
+                        )
+                    )
+
+                # instantiate template
+                bb = bb(**operation["kwargs"])
+
+            else:
+                # operation is not a template
+                if bb.is_template():
+                    raise ValueError(
+                        "Included operation {} missing keyword arguments {}".format(
+                            operation["op"], bb.parameters
+                        )
+                    )
+
+            # mode mapping dictionary
+            mode_map = dict(zip(bb.modes, operation["modes"]))
```

Comment:
I can see what the intent is here, but I can't figure out what is happening. 🤔 
The modes in the included program are mapped to the modes of the including program, or the other way around?

Reply:
The former; maps the modes of the included program to the modes it is applied to in the including program. I've updated the comment with this info.